api/users_test.go

```diff
@@ -380,6 +380,156 @@ func TestGetUserHandler(t *testing.T) {
+					user := &cognitoidentityprovider.AdminGetUserOutput{
+						UserAttributes: successfullyGetUser,
+						UserStatus:     &status,
+						Username:       &userId,
+					}
+					return user, nil
+				},
+				forename,
+				http.StatusInternalServerError,
+			},
+			{
+				//404 response from Cognito
+				func(userInput *cognitoidentityprovider.AdminUpdateUserAttributesInput) (*cognitoidentityprovider.AdminUpdateUserAttributesOutput, error) {
+					awsErrCode := "UserNotFoundException"
+					awsErrMessage := "user could not be found"
+					awsOrigErr := errors.New(awsErrCode)
+					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
+					return nil, awsErr
+				},
+				func(userInput *cognitoidentityprovider.AdminGetUserInput) (*cognitoidentityprovider.AdminGetUserOutput, error) {
+					user := &cognitoidentityprovider.AdminGetUserOutput{
+						UserAttributes: successfullyGetUser,
+						UserStatus:     &status,
+						Username:       &userId,
+					}
+					return user, nil
+				},
+				forename,
+				http.StatusNotFound,
+			},
+			{
+				//local validation failure
+				func(userInput *cognitoidentityprovider.AdminUpdateUserAttributesInput) (*cognitoidentityprovider.AdminUpdateUserAttributesOutput, error) {
+					user := &cognitoidentityprovider.AdminUpdateUserAttributesOutput{}
+					return user, nil
+				},
+				func(userInput *cognitoidentityprovider.AdminGetUserInput) (*cognitoidentityprovider.AdminGetUserOutput, error) {
+					user := &cognitoidentityprovider.AdminGetUserOutput{
+						UserAttributes: successfullyGetUser,
+						UserStatus:     &status,
+						Username:       &userId,
+					}
+					return user, nil
+				},
+				"",
+				http.StatusBadRequest,
+			},
+			{
+				//reload user details failure
+				func(userInput *cognitoidentityprovider.AdminUpdateUserAttributesInput) (*cognitoidentityprovider.AdminUpdateUserAttributesOutput, error) {
+					user := &cognitoidentityprovider.AdminUpdateUserAttributesOutput{}
+					return user, nil
+				},
+				func(userInput *cognitoidentityprovider.AdminGetUserInput) (*cognitoidentityprovider.AdminGetUserOutput, error) {
+					awsErrCode := "UserNotFoundException"
+					awsErrMessage := "user could not be found"
+					awsOrigErr := errors.New(awsErrCode)
+					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
+					return nil, awsErr
+				},
+				forename,
+				http.StatusInternalServerError,
+			},
+		}
+
+		for _, tt := range adminCreateUsersTests {
+			m.AdminUpdateUserAttributesFunc = tt.updateUserFunction
+			m.AdminGetUserFunc = tt.getUserFunction
+
+			postBody := map[string]interface{}{"forename": tt.userForename, "lastname": lastname, "status": status, "role_type": roleType}
+			body, _ := json.Marshal(postBody)
+			r := httptest.NewRequest(http.MethodGet, userEndPoint, bytes.NewReader(body))
+
+			successResponse, errorResponse := api.UpdateUserHandler(ctx, w, r)
+
+			// Check whether testing a success or error case
+			if tt.httpResponse > 399 {
```

Comment:
this is relying on the tests returning the right responses for each test case. Is there a way we can add the "expectation" to the test case struct? That way if the test returns an incorrect response the expectation defined in the test case will ensure the test fails appropriately.

Reply:
This was actually relying on response code a set in the test case struct to determine which set of assertions to user for the test case.
But I agree having the assertions as part of the test case struct makes it cleaner and easier to follow.